### PR TITLE
:sparkles: [#1007] Configurable IOType per FileField

### PR DESCRIPTION
**Pull request by stevenbal**
Fixes: #1007

Changes:
- Added the option to select an informatieobjecttype per FileField for ZGW APIs and Objects API registration backends

### CODE REVIEW

**Comment by stevenbal** [comment]
To start, I added the catalogus.domein to the labels (so you can at least see to which catalogus the informatieobjecttype belongs to). I was looking into using optgroups, but it seems like the FormIO implementation of `select` doesn't support this?

Screenshot from 2022-01-07 15-56-20

**Comment by stevenbal** [comment]
I guess to allow `optgroups`, we would have to override the template for `select`? I'm not sure how to do this though

**Comment by sergei-maertens** [comment]
@stevenbal right, it looks like Formio does not support that out of the box looking at https://github.com/formio/formio.js/blob/4.12.x/src/templates/bootstrap/selectOption/form.ejs :(

Might be an option to define a custom `Select` component subclass, which can render a custom `SelectOption` template. Feel free to work on this, but I probably won't be able to review this/follow up until after the demo coming Monday.

**Comment by stevenbal** [comment]
> edit: did a quick check, and choices.js (used by Formio for selects) does support optgroup, so maybe with some configuration it might be possible to beat it into submission.

I tried to see if I could somehow get this to work, but I don't know where I would change this. I thought I would be able to override `templates/selectOption.js`, but `select.js` doesn't seem to use this template.

I guess I'll leave this out for now?

**Comment by sergei-maertens** [comment]
Let's park it for now, since I don't have bandwith at the moment to help with this.

**Comment by stevenbal** [comment]
I got optgroups working, the only thing is that they do not show up when performing search
cbzxFxtJ4h

**Comment by sergei-maertens** [comment]
@stevenbal the black highlight was fixed last friday! :)

**Comment by stevenbal** [comment]
> for the optgroups - if the behaviour is consistent with https://formio.github.io/formio.js/app/builder, then it is what it is

By default formio doesn't seem to support optgroups (from what I see), but search works the same as in the builder demo

File: src/openforms/registrations/contrib/zgw_apis/service.py [hunk 1]
```diff
@@ -142,6 +143,28 @@ def create_csv_document(
     return create_document(name, base64_body, options, get_drc=get_drc)
 
 
+def retrieve_attachment_iotype(
+    submission_attachment: SubmissionFileAttachment, default_iotype: str
+) -> str:
```
Comment (sergei-maertens) [inline src/openforms/registrations/contrib/zgw_apis/service.py:148]: The function name is confusing, the `retrieve` implies to me that an HTTP call is made, but that's not the case.

Wouldn't it make more sense to make this a method on the `SubmissionFileAttachment` too? It basically should only interface with `openforms.formio.service` and internal data structures.

**Review by sergei-maertens — CHANGES_REQUESTED** [review]

File: src/openforms/registrations/api/views.py [hunk 2]
```diff
@@ -50,3 +54,62 @@ def get_objects(self):
...
+                )
+                for iotype in iotypen_data["results"]
+            ]
+
+            if iotypen_data["next"]:
+                url = furl(iotypen_data["next"])
+                query_params = url.query.params
+            else:
+                break
+        return iotypen
```
[the remark this answers is not in the document]
Comment (stevenbal) [reply src/openforms/registrations/api/views.py:115]: Hmm, I tried an approach like this to fetch the IOtypen in parallel:

```python
def get_objects(self):
        ...

        catalogus_data = get_paginated_results(client, "catalogus")
        catalogus_mapping = {
            catalogus["url"]: factory(Catalogus, catalogus)
            for catalogus in catalogus_data
        }

        def get_iotypen(catalogus: str) -> List[dict]:
            iotypen_data = get_paginated_results(
                client, "informatieobjecttype", query_params={"catalogus": catalogus}
            )
            iotypen = [
                {
                    "informatieobjecttype": factory(InformatieObjectType, iotype),
                    "catalogus": catalogus_mapping[iotype["catalogus"]],
                }
                for iotype in iotypen_data
            ]
            return iotypen

        with parallel() as executor:
            _data = executor.map(get_iotypen, list(catalogus_mapping.keys()))

        iotypen = sum(list(_data), [])
        return iotypen
```

But this actually seemed to slow things down slightly, is it worth it here to fetch stuff in parallel :thinking:?
Screenshot from 2022-05-30 12-46-00

Screenshot from 2022-05-30 12-46-18

[the remark this answers is not in the document]
Comment (sergei-maertens) [reply src/openforms/registrations/api/views.py]: Numbers don't lie - current case is probably the more efficient one then. Probably it'll only start making a difference if there are many large catalogi (involving pagination and multiple sequential calls), as long as it all fits in a single call you don't have the HTTP overhead. So okay to leave it as is now, thanks for profiling it!

**Review by sergei-maertens — APPROVED** [review]

**Pull request merged**